### PR TITLE
Auto-generate SDK API reference

### PR DESCRIPTION
**Pull request by jhkennedy**
## Primarily: 
* audodoc SDK API 

### Justifcation
auto-generating an API reference for the SDK provides:
* modules, functions, classes, etc are all self documenting
  * typehints and docstrings are used to generate the documentation
  * updating the documentation happens simultaneously with code updates

*Note: All auto-documentation packages for Sphinx of MkDocs require that anything auto-documented is installed/importable.*

### Approaches
There are three ways to approach auto-documentation of external repositories as described below, with some reasoning as to why I took this approach. 

#### Publish SDK docs on its own and link directly to it 

We could publish the SDK docs to a GitHub Pages site associated with the SKD repository (would live at https://ASFHyP3.github.io/hyp3-sdk) an link directly to it in this website

##### Advantages:
* no incorporating or depending on any externals
* no dependency management here (all handled in the SDK repo)

##### Disadvantages:
* our documentation navigation structure is constrained by this organizations repository structure
* multiple websites for documentation mean navigating between this organizations parts is harder (no single table of contents and navigation menu)
* SEO optimization is harder with multiple sites and placing it inside ASFs website (eventual goal) will be harder

#### use [mkdocs-monorepo-plugin](https://github.com/spotify/mkdocs-monorepo-plugin) to include entire documentation site

Instead of linking directly to completely separate website, we could consume the MkDocs structure for the SDK and include it in its own section here.

##### Advantages
* single website is generated for entire organization
  * easy navigation between organization parts
  * easy deployment inside ASF's website and easier to optimize SEO
* website structure is *not* constrained by organizations repository structure

##### Disadvantages
* the SDK must be included as a [git submodule](https://git-scm.com/docs/git-submodule)
* the generating environment must have the SDK installed (and all its dependencies)
* **The SDK docs will *only* appear under a section (i.e., navigation would look like `HyP3 -> Using HyP3 -> SDK -> hyp3-sdk`) even for single-page documentation of the SDK**

#### (this) [mkdocs-gitsnippet-plugin](https://github.com/Jlrine2/mkdocs-gitsnippet-plugin) a single-page docs site that auto-generates the API spec

Instead of incorporating the entire docs structure, we instead directly consume a single markdown file containing all of the SDK documentation, like we are currently doing for the ArcGIS Toolbox

##### Advantages
* single website is generated for entire organization
  * easy navigation between organization parts
  * easy deployment inside ASF's website and easier to optimize SEO
* website structure is *not* constrained by organizations repository structure

##### Disadvantages
* the generating environment must have the SDK installed (and all its dependencies)
* The SDK docs must be a single page (but can be placed anyway we like in the navigation structure)

#### Deciding factor

Because the SDK is small and fits okay in a single-page docs site, `mkdocs-gitsnippet-plugin` appears the way to go. The `mkdocs-monorepo-plugin` works great for docs site that need to have multiple pages but requires using submodules and will need to be added in such a way a containing section doesn't become too burdensome. 

## Also:
* drops sphinx; mkdocstrings works great to autodocs and so we won't need sphinx
* GitHub Actions:
  * add manual trigger to deploy workflow (to regenerate docs with release of GIS-tools) 
  * add manual bump SDK version workflow (to regenerate SDK API reference with new release)
  
  *Note: manual workflows can be triggered by a POST to the github REST API*

## Requires:
- [x] Release of HyP3 SDK v0.2.2 or greater (should update `conda-env.yml` and `docs/using/sdk.md` with specific tag)

### CODE REVIEW

File: .github/workflows/deploy_to_github_io.yml [hunk 1]
```diff
@@ -4,11 +4,13 @@ on:
   push:
     branches:
       - main
+  workflow_dispatch:
```
Comment (asjohnston-asf) [inline .github/workflows/deploy_to_github_io.yml:7]: is this yaml key missing it's content?  not sure if this is valid yaml, or if it 'just works', https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/ seems to suggest you might not have any inputs or tags

Comment (jhkennedy) [reply .github/workflows/deploy_to_github_io.yml:7]: > You must append a colon (:) to all events, including events without configuration.

https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-using-multiple-events-with-activity-types-or-configuration

**Review by asjohnston-asf — APPROVED** [review]
Looks reasonable, although I'll want to watch each of the new deploy actions execute before I trust them

**Comment by jhkennedy** [comment]
> Looks reasonable, although I'll want to watch each of the new deploy actions execute before I trust them

Definitely -- me too 100%.

**Pull request merged**